src/com/ms/silverking/net/async/PersistentAsyncServer.java

```diff
@@ -8,402 +8,453 @@
-        } catch (UnknownHostException uhe) {
-            throw new RuntimeException(uhe);
-        }
+    return connection;
+  }
+
+  private Connection getConnectionSlow(InetSocketAddress dest, long deadline, String context) throws ConnectException {
+    Connection connection;
+    ReentrantLock destNewConnectionLock;
+
+    destNewConnectionLock = newConnectionLocks.get(dest);
+    if (destNewConnectionLock == null) {
+      newConnectionLocks.putIfAbsent(dest, new ReentrantLock());
+      destNewConnectionLock = newConnectionLocks.get(dest);
+    }
+    destNewConnectionLock.lock();
+    try {
+      connection = connections.get(dest);
+      if (connection == null) {
+        connection = createConnection(dest, deadline, context);
+      }
+      return connection;
+    } finally {
+      destNewConnectionLock.unlock();
+    }
+  }
+
+  /**
+   * only called when the given connection does not exist
+   *
+   * @return
+   */
+  private Connection createConnection(InetSocketAddress dest, long deadline, String context) throws ConnectException {
+    RandomBackoff backoff;
+    IPAndPort _dest;
+
+    if (Log.levelMet(Level.INFO)) {
+      String tag = ((isClient) ? "[Client]" : "") + "createConnection: ";
+      Log.info(tag, dest);
     }
-    
-    private Connection getConnectionFast(InetSocketAddress dest, long deadline) throws ConnectException {
-        Connection    connection;
-        
-        connection = connections.get(dest);
-        if (connection == null) {
-            connection = getConnectionSlow(dest, deadline);
+
+    _dest = new IPAndPort(dest);
+
+    deadline = Math.min(deadline,
+        SystemTimeUtil.skSystemTimeSource.absTimeMillis() + newConnectionTimeoutController.getMaxRelativeTimeoutMillis(
+            _dest));
+
+    backoff = null;
+    while (isRunning) {
+      if (addressStatusProvider != null && !addressStatusProvider.isAddressStatusProviderThread(context)
+          && !addressStatusProvider.isHealthy(dest)) {
+        throw new UnhealthyConnectionAttemptException("Connection attempted to unhealthy address: " + dest);
+      }
+
+      try {
+        T connection;
+
+        connection = asyncServer.newOutgoingConnection(dest, this);
+        connections.putIfAbsent(dest, connection);
+        if (suspectAddressListener != null) {
+          suspectAddressListener.removeSuspect(dest);
         }
         return connection;
-    }
-    
-    private Connection getConnectionSlow(InetSocketAddress dest, long deadline) throws ConnectException {
-        Connection        connection;
-        ReentrantLock    destNewConnectionLock;
-        
-        destNewConnectionLock = newConnectionLocks.get(dest);
-        if (destNewConnectionLock == null) {
-            newConnectionLocks.putIfAbsent(dest, new ReentrantLock());
-            destNewConnectionLock = newConnectionLocks.get(dest);
+      } catch (ConnectionAbsorbException cae) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 379, 400.